From 37bf10442081a2a1ce338c972307666eb8e186b6 Mon Sep 17 00:00:00 2001
From: Thomas Tai <Thomas.Tai@windriver.com>
Date: Tue, 14 Oct 2008 16:37:07 -0400
Subject: [PATCH 1/3] Correct timer frequency

A hardware timer is available in tsi108 bridge, the default
timer frequency is 133Mhz and the default prescaler is 1. This
hardware timer is used in hwtimer API.

Signed-off-by: Thomas Tai <Thomas.Tai@windriver.com>
---
 arch/powerpc/sysdev/mpic.c |   25 +++++++++++++++++++++++++
 1 files changed, 25 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/sysdev/mpic.c b/arch/powerpc/sysdev/mpic.c
index 51612ae..e7de3b6 100644
--- a/arch/powerpc/sysdev/mpic.c
+++ b/arch/powerpc/sysdev/mpic.c
@@ -269,15 +269,24 @@ static inline void _mpic_irq_write(struct mpic *mpic, unsigned int src_no,
 #include <linux/hwtimer.h>
 #include <linux/delay.h>
 
+#ifdef CONFIG_MPC7448HPC2
+#define TIMER_DEFAULT_DIVISOR 1
+#else
 #define TIMER_DEFAULT_DIVISOR 8
+#endif
 #ifdef CONFIG_MPC85xx
 static int pll_divisor[4]={1,2,1,2};
 static int pll_dividend[4]={2,5,3,7};
 static int pll_index=1;
 #define PROC_PLL_RATIO pll_divisor[pll_index]/pll_dividend[pll_index]
 #else
+#ifdef CONFIG_MPC7448HPC2
+/* timer is driven by the 133Mhz fabric clock */
+#define TIMER_DEFAULT_FREQ (133*1000000)
+#else
 #define PROC_PLL_RATIO 2/5
 #endif
+#endif
 
 static unsigned long mpic_hwtimer_freq = 0;
 
@@ -300,10 +309,18 @@ static int mpic_hwtimer_set_freq(int freq)
 	if ((freq < mpic_hwtimer_data.min_freq) || (freq > mpic_hwtimer_data.max_freq))
 		return -EINVAL;
 	mpic_hwtimer_freq = freq;
+#ifdef CONFIG_MPC7448HPC2
+	/* timer is driven by the 133Mhz fabric clock */
+	mpic_write(mpic->tmregs,
+		   mpic_hwtimer_index * MPIC_INFO(TIMER_STRIDE) +
+		   MPIC_INFO(TIMER_BASE_CNT),
+		   TIMER_DEFAULT_FREQ / TIMER_DEFAULT_DIVISOR / mpic_hwtimer_freq);
+#else
 	mpic_write(mpic->tmregs,
 		   mpic_hwtimer_index * MPIC_INFO(TIMER_STRIDE) +
 		   MPIC_INFO(TIMER_BASE_CNT),
 		   ppc_proc_freq * PROC_PLL_RATIO / TIMER_DEFAULT_DIVISOR / mpic_hwtimer_freq);
+#endif
 	return 0;
 }
 
@@ -330,10 +347,18 @@ static int mpic_hwtimer_start(void)
 		   1);
 
 	/* Set ticks. */
+#ifdef CONFIG_MPC7448HPC2
+	/* timer is driven by the 133Mhz fabric clock */
+	mpic_write(mpic->tmregs,
+		   mpic_hwtimer_index * MPIC_INFO(TIMER_STRIDE) +
+		   MPIC_INFO(TIMER_BASE_CNT),
+		   TIMER_DEFAULT_FREQ / TIMER_DEFAULT_DIVISOR / mpic_hwtimer_freq);
+#else
 	mpic_write(mpic->tmregs,
 		   mpic_hwtimer_index * MPIC_INFO(TIMER_STRIDE) +
 		   MPIC_INFO(TIMER_BASE_CNT),
 		   ppc_proc_freq * PROC_PLL_RATIO / TIMER_DEFAULT_DIVISOR / mpic_hwtimer_freq);
+#endif
 	tmp = mpic_read(mpic->tmregs, mpic_hwtimer_index * MPIC_INFO(TIMER_STRIDE) +
 		   	MPIC_INFO(TIMER_VECTOR_PRI));
 
-- 
1.6.0

